File: novachat-common/src/main/java/com/nova/chat/common/protocol/VarInt.java
package com.nova.chat.common.protocol;

import io.netty.buffer.ByteBuf;

/**
 * VarInt encoder/decoder for NovaProtocol.
 * VarInt is a variable-length integer encoding that uses 1-5 bytes.
 * Each byte uses 7 bits for data and 1 bit (MSB) as continuation flag.
 */
public final class VarInt {

    private static final int SEGMENT_BITS = 0x7F;
    private static final int CONTINUE_BIT = 0x80;
    private static final int MAX_VARINT_SIZE = 5;

    private VarInt() {
        // Utility class
    }

    /**
     * Reads a VarInt from the given ByteBuf.
     *
     * @param buf the buffer to read from
     * @return the decoded integer value
     * @throws IllegalArgumentException if the VarInt is too large (more than 5 bytes)
     */
    public static int readVarInt(ByteBuf buf) {
        int value = 0;
        int position = 0;
        byte currentByte;

        while (true) {
            currentByte = buf.readByte();
            value |= (currentByte & SEGMENT_BITS) << position;

            if ((currentByte & CONTINUE_BIT) == 0) {
                break;
            }

            position += 7;

            if (position >= 32) {
                throw new IllegalArgumentException("VarInt is too big");
            }
        }

        return value;
    }

    /**
     * Writes a VarInt to the given ByteBuf.
     *
     * @param buf   the buffer to write to
     * @param value the integer value to encode
     */
    public static void writeVarInt(ByteBuf buf, int value) {
        while (true) {
            if ((value & ~SEGMENT_BITS) == 0) {
                buf.writeByte(value);
                return;
            }

            buf.writeByte((value & SEGMENT_BITS) | CONTINUE_BIT);
            value >>>= 7;
        }
    }


    /**
     * Calculates the number of bytes needed to encode the given value as a VarInt.
     *
     * @param value the integer value
     * @return the number of bytes needed (1-5)
     */
    public static int getVarIntSize(int value) {
        if ((value & (0xFFFFFFFF << 7)) == 0) {
            return 1;
        }
        if ((value & (0xFFFFFFFF << 14)) == 0) {
            return 2;
        }
        if ((value & (0xFFFFFFFF << 21)) == 0) {
            return 3;
        }
        if ((value & (0xFFFFFFFF << 28)) == 0) {
            return 4;
        }
        return 5;
    }

    /**
     * Encodes an integer value to a byte array as VarInt.
     *
     * @param value the integer value to encode
     * @return the encoded byte array
     */
    public static byte[] encode(int value) {
        byte[] result = new byte[getVarIntSize(value)];
        int index = 0;

        while (true) {
            if ((value & ~SEGMENT_BITS) == 0) {
                result[index] = (byte) value;
                return result;
            }

            result[index++] = (byte) ((value & SEGMENT_BITS) | CONTINUE_BIT);
            value >>>= 7;
        }
    }

    /**
     * Decodes a VarInt from a byte array.
     *
     * @param bytes the byte array to decode from
     * @return the decoded integer value
     * @throws IllegalArgumentException if the VarInt is too large
     */
    public static int decode(byte[] bytes) {
        int value = 0;
        int position = 0;

        for (byte currentByte : bytes) {
            value |= (currentByte & SEGMENT_BITS) << position;

            if ((currentByte & CONTINUE_BIT) == 0) {
                break;
            }

            position += 7;

            if (position >= 32) {
                throw new IllegalArgumentException("VarInt is too big");
            }
        }

        return value;
    }
}

File: NovaChat/Bedrock/pmmp/src/NovaChat/Protocol/VarInt.php
<?php

declare(strict_types=1);

namespace NovaChat\Protocol;

use InvalidArgumentException;

/**
 * VarInt encoder/decoder for NovaProtocol.
 * 
 * VarInt is a variable-length integer encoding used in the NovaProtocol
 * to efficiently encode integers of varying sizes.
 * 
 * Requirements:
 * - 9.1: THE NovaChat-PMMP SHALL 实现 VarInt 编解码器
 */
class VarInt {
    
    /** Maximum bytes for a VarInt (5 bytes for 32-bit integers) */
    public const MAX_BYTES = 5;
    
    /** Maximum value for a VarInt */
    public const MAX_VALUE = 2147483647;
    
    /** Minimum value for a VarInt */
    public const MIN_VALUE = -2147483648;
    
    /**
     * Encodes an integer value to VarInt bytes.
     * 
     * @param int $value The integer value to encode
     * @return string The encoded VarInt bytes
     * @throws InvalidArgumentException If the value is out of range
     */
    public static function write(int $value): string {
        if ($value > self::MAX_VALUE || $value < self::MIN_VALUE) {
            throw new InvalidArgumentException("Value out of VarInt range: $value");
        }
        
        $result = "";
        $unsigned = $value & 0xFFFFFFFF;
        
        while (true) {
            if (($unsigned & ~0x7F) === 0) {
                $result .= chr($unsigned);
                break;
            }
            $result .= chr(($unsigned & 0x7F) | 0x80);
            $unsigned >>= 7;
        }
        
        return $result;
    }
    
    /**
     * Reads a VarInt from a buffer at the specified offset.
     * 
     * @param string $buffer The buffer to read from
     * @param int &$offset The offset to start reading from (will be updated)
     * @return int The decoded integer value
     * @throws InvalidArgumentException If the VarInt is malformed
     */
    public static function read(string $buffer, int &$offset): int {
        $result = 0;
        $shift = 0;

        while (true) {
            if ($offset >= strlen($buffer)) {
                throw new InvalidArgumentException("Buffer underflow while reading VarInt");
            }

            // Check for overflow BEFORE reading more bytes. After the 5th byte
            // with its continuation bit set, $shift becomes 35 (>= 32), so we
            // reject strictly more than 5 bytes without consuming a 6th byte.
            if ($shift >= 32) {
                throw new InvalidArgumentException("VarInt is too big");
            }

            $byte = ord($buffer[$offset++]);

            // On the 5th byte ($shift === 28), only the low 4 bits are valid for
            // a 32-bit integer. Mask with 0x0F to reject bits that would shift
            // past bit 31 and silently corrupt the result.
            if ($shift === 28) {
                $result |= ($byte & 0x0F) << $shift;
            } else {
                $result |= ($byte & 0x7F) << $shift;
            }

            if (($byte & 0x80) === 0) {
                break;
            }

            $shift += 7;
        }

        // Convert to signed 32-bit integer
        if ($result > 0x7FFFFFFF) {
            $result -= 0x100000000;
        }

        return $result;
    }
    
    /**
     * Calculates the size in bytes of a VarInt-encoded value.
     * 
     * @param int $value The value to calculate size for
     * @return int The number of bytes needed to encode the value
     */
    public static function size(int $value): int {
        $unsigned = $value & 0xFFFFFFFF;
        $size = 0;
        
        do {
            $size++;
            $unsigned >>= 7;
        } while ($unsigned !== 0);
        
        return $size;
    }
}
